File: 90-subsets-ii/subsets-ii.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> curr;
        sort(nums.begin(), nums.end());
        dfs(curr, 0, nums);
        return res;
    }
    void dfs(vector<int>& curr, int i, vector<int>& nums) {
        if (i>=nums.size()) {
            res.push_back(curr);
            return;
        }

        curr.push_back(nums[i]);
        dfs(curr, i+1, nums);

        curr.pop_back();
        while (i+1<nums.size() && nums[i] == nums[i+1]) i+=1;
        dfs(curr, i+1, nums);
    }
};

File: 90-subsets-ii/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []

        nums.sort()

        def backtrack(i, curr):
            # base
            if i>= len(nums):
                ans.append(curr[:])
                return
            
            # main
            # taken
            curr.append(nums[i])
            backtrack(i+1, curr)

            # not taken
            curr.remove(nums[i])
            while i+1 < len(nums) and nums[i] == nums[i+1]:
                i+=1
            backtrack(i+1, curr)

        backtrack(0, [])

        return ans